File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        int open = n;
        int closed = n;
        ArrayList<String> a = new ArrayList<String>();
        birthday(open,closed,"",a);
        return a;
    }

    public void birthday(int open,int closed,String s , ArrayList<String>res){
        if(open == closed && open == 0){
            res.add(s);
            return;
        }
        if (open > 0) {
            birthday(open-1, closed, s + "(", res);
        }

        if (closed > open) {
            birthday(open, closed - 1, s + ")",  res);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def rec(i,oc,cc,c):
            if i == 2*n:
                ans.append(c)
                return
            if cc > oc:
                return
            if oc < n:
                rec(i+1,oc+1,cc,c+"(")
            if cc < oc:
                rec(i+1,oc,cc+1,c+')')

        ans = list()
        rec(0,0,0,"")
        return ans